### DISCUSSION TITLE
[Ongoing] New API Ideas

### DISCUSSION THREAD

**Discussion by lokka30** [ArcanePlugins/Treasury#128] [Ideas]
# New API Ideas!

Hey there everyone, if you have any ideas for new APIs to go into Treasury, feel free to chat below! If it is regarding an existing API idea, please reply to it rather than create a new comment.

# Q/A

## What APIs won't be added?

* Anything that can't be scaled to multiple platforms (including proxies)

* Anything handling Minecraft internals (NMS)
  * Yes, including packets

* Utility APIs such as database management. Treasury is a library for integrations, not utilities.

## How can I be alerted when any of these APIs are released?

Simply 'Watch Resource' on Treasury's SpigotMC page and see the top of the changelogs. We rarely update so you won't get spammed. :)

**Comment by lokka30** [comment]
# Permissions+Chat API

## Likelihood:
🟢

## Features:
* `?`

## Code ideas:
* `?`

**Reply by MrIvanPlays** [reply]
We can combine chat api with permissions api. chat api is basically prefixes. We can go with luckperms' approach by setting prefixes with permissions.

**Reply by lokka30** [reply]
I think it's best we include the authors of chat plugins in this conversation, they'd know best what they're looking for. :)
It's possible we could merge the APIs though again I'd rather not say anything since I don't have enough experience in this specific field. Same goes with the Permissions one.

**Reply by MrIvanPlays** [reply]
I have experience with it. Combining them would mean 1 better api with enough content rather than 2 apis with not-so-much content.

**Reply by Jikoo** [reply]
I agree, I think that chat parsing is best left to \<library of choice\>. I wouldn't hate seeing a library for a better AsyncPlayerChatEvent that uses components, but I'm not sure if that's even useful since most platforms seem to be migrating to Adventure. Pretty sure Paper already offers that in the base event at this point. Chat prefixes and whatnot seem good, but should likely be attached to permissions.

I guess the question for me with this is "do we want to expose and expand Treasury's language system to allow other plugins to piggyback it" because that would give us an actual reason for creating an API for things involving chat.

**Reply by lokka30** [reply]
Note: hidden the individual `Permissions` API post, and edited this original one to include `Permissions`. In other words, they're merged.

**Reply by MrNemo64** [reply]
> We can combine chat api with permissions api. chat api is basically prefixes. We can go with luckperms' approach by setting prefixes with permissions.

In my opinion a Chat API could do much more than this. See [here](https://github.com/lokka30/Treasury/issues/14#issuecomment-1013656064) my ideas

**Reply by MrIvanPlays** [reply]
your idea is already present in adventure, which we shall support OOB, as it has gained adoption in sponge, velocity and paper.

**Reply by lokka30** [reply]
^ translation: everything except spigot/bungee. 😄

**Reply by lokka30** [reply]
Moving @Jannyboy11's post here:

I'd be happy if the Treasury Permissions API will have the same advantages as its economy API over Vault.

More concretely, these features are on my wishlist:
- The api must be able to answer whether offline players have a permission, through the use of UUIDs.
- Must be able to be consumed from async code and/or return a result outside the server's primary thread.
- I'd like all variants of "hasPermission" to return a kind of TriState instead of a boolean. e.g. `SPECIFIED_TRUE`, `SPECIFIED_FALSE` and `UNSPECIFIED`.

Personally I only care about an API to get permission info for players, but I can imagine others needing support for permission groups too.

(Just to reiterate, this is from Jan, not me).

**Reply by MrIvanPlays** [reply]
We already have a TriState implementation so that's been cleared out.

**Reply by lokka30** [reply]
I think the only thing we need to note down for permissions at least is the ability to retrieve it whilst the user is offline of course. Should be very easy for permission plugins to handle. My only concern is of the permission plugins that are conditional, e.g. you have this permission but only in this world. So maybe the offline permissions need to be an optional feature. This will require more discussion it seems.

**Reply by MrIvanPlays** [reply]
We shall not be pushing hard on new APIs. Awfully lot of discussion is needed in terms of how they shall be implemented and etc. Such discussions cannot happen in a day.

**Reply by lokka30** [reply]
Of course. For such an important API especially, this may take a while to design and program.

**Reply by A248** [reply]
To revive this discussion, here are some suggestions:

* Regarding conditional permissions, the approach used by the SpongeAPI and LuckPerms utilizes key-value pairs called "contexts." Treasury's API, accordingly, could accept a `Set<PermissionContext>` or a `Map<String, String>` when a permission is requested for a possibly offline player. For example, this map might be `Map.of("world", "world_the_end", "server", "lobby")` to indicate the world *world_the_end* and the backend server *lobby*.
* There should be also be a method, for online players only, which checks for a permission node using the player's currently active contexts.
* Another aspect of the permissions API should be the ability to request permission data for a possibly offline user before performing any permission checks. This way, the caller may indicate a desire to have the data loaded in-memory before performing a specific query.

On Sponge, if or rather when Treasury is implemented for that platform, there should be a default implementation of the Treasury permissions API which delegates to Sponge's PermissionService.

**Comment by lokka30** [comment]
Permissions API

**Comment by lokka30** [comment]
# Land Protection API

## Likelihood:
🟡 Partial

## Features:
* Region Permissions enum states what a player can do in a region (claim).
* Check what Region Permissions a player has for a specific location.
* Get the names of the plugins that affect a player's Region Permissions for a specific location. Allows plugins to be configured e.g. 'ignore worldguard regions for random teleport'.

## Code ideas:
* `?`

**Reply by MrIvanPlays** [reply]
Why?

**Reply by lokka30** [reply]
This actually has several use cases. For example, let's take a Random Teleport plugin. You don't want to be spawned randomly in the middle of someone's claim, right? To achieve this, a Random Teleport plugin will need to hopefully have a direct integration with the land protection plugin the server is using. Could be something popular like GriefPrevention, GriefDefender, Factions, Towny etc. But what if it's not a popular plugin? It's very unfeasible to maintain so many integrations just to cover most use cases. Rather, a central API would be so much better.
This is absolutely not limited to random teleport plugins. I can name many other things. Pet plugins. Fly-on-condition plugins, e.g. only fly in your claim. Custom enchantment plugins. The list goes on

**Reply by MrIvanPlays** [reply]
This won't offer that much content as I can see from your reply. But now I agree that we need such thing.

**Reply by Jikoo** [reply]
Re: RTP - generally you'd just fire a fake BlockBreakEvent to check all anti-grief plugins for build rights, but I definitely agree that more specific use cases would be nice. Goodness knows I'm tired of [offering support for any and all anti-grief plugins](https://github.com/Jikoo/Regionerator/tree/8f07e292cd3cd9930458edf6d82b025f4d326ff4/src/main/java/com/github/jikoo/regionerator/hooks).

**Reply by lokka30** [reply]
I don't see why 'content' is an issue - our APIs can be simple yet highly effective. I also believe this is the case for the proposed Vanish API. I can elaborate on that if you like under that post.

The first method I can think of is obviously something like `hasRegionPermission(player, RegionPermission.BREAK_BLOCKS, location, new RegionSubscriber<TriState>(subscription -> {...}));`.

**Reply by lokka30** [reply]
> Re: RTP - generally you'd just fire a fake BlockBreakEvent

This is definitely a 'solution', but it is obviously not ideal, and could cause issues with plugins that listen to BlockBreakEvent too. My hope is that Treasury might be able to solve this. 😉

**Reply by MrIvanPlays** [reply]
> `new RegionSubscriber<Boolean>(subscription -> {...}))`

`RegionSubscriber<TriState>`*

Content is an issue because why should we have a unified api if it has no more than 3 methods?

**Reply by lokka30** [reply]
> > `new RegionSubscriber<Boolean>(subscription -> {...}))`
> 
> `RegionSubscriber<TriState>`*

Great point 😄

> Content is an issue because why should we have a unified api if it has no more than 3 methods?

If it only takes 3 methods to solve a frustrating issue, then all the better I think.

**Reply by MrNemo64** [reply]
> Content is an issue because why should we have a unified api if it has no more than 3 methods?

I think this API could have more than 3 methods. My ideas are this:
The `LandProtector` interface. This is what would be implemented by providers and has methods like `void canPerformAction(Entiry, Action, Position, LandProtectorSubscriber<TriState>)`, `void reteiveRegisteredAreas(LandProtectorSubscriber<Collection<ProtectedArea>>)`, `Set<OptionalLandProtectionFeatures> getSupportedOptionalLandProtectionApiFeatures()` etc
The `ProtectedArea` interface would represent a protected area/zone in the world and it's somewhat inspired by the [Shape](https://docs.oracle.com/javase/8/docs/api/java/awt/Shape.html) class. This interface would define methods like `void canPerformAction(Entity, Action, LandProtectorSubscriber<TriState>)`, `void containsPosition(Position, LandProtectorSubscriber<Boolean>)`, `void intersects(Cube, LandProtectorSubscriber<Boolean>)` `void getPathIterator(LandProtectorSubscriber<Iterator<Segment>>)`, `void getBlockIterator(LandProtectorSubscriber<Iterator<Block>>)` *(dangerous as interactions with blocks normaly can only be performed onthe main thread)* and maybe some more

**Comment by lokka30** [comment]
# Vanish API

## Likelihood:

🌕 Partial

## Features:

* Check if an online/offline player can see another online/offline player (vanished state)

## Code ideas:

* `record VanishedState(UUID playerId, Map<UUID otherPlayerId, TriState visibility)`
* `retrieveVanishedState(UUID playerId, VanishSubscriber<VanishedState> subscriber);`
* `retrieveVanishedState(UUID playerId, UUID targetId, VanishSubscriber<TriState> subscriber);`]

Note: `TriState`s shall be used because:
* `TriState.TRUE` = the viewer is said to **must** be able to see the target
* `TriState.UNSPECIFIED` = the default value: the viewer can see the target but this can change
* `TriState.FALSE` = the viewer can't see the target

**Reply by MrNemo64** [reply]
Just to check if a player is in vanish mode? Seems laking of content
Maybe a modderator API (not sure what to call it) that gives the vanish API and the already mentioned idea of
> Ideally since there are plugins that store the interactions with a block(X played Y, C broke, A clicked it) I'd like to be able to get that information

**Reply by lokka30** [reply]
Added as a separate post in this discussion :+1:

but yeah - there is no central vanish API at the moment.

**Reply by MrIvanPlays** [reply]
As MrNemo said vanish api does not have any content. not needed at all

**Reply by lokka30** [reply]
It's still on my agenda but I don't mind which direction this idea goes.

**Reply by MrIvanPlays** [reply]
another concern I have ; which plugins are going to use such api? AFAIK the only plugins which care whether a player is vanished or not is the vanish plugin itself.

**Reply by Jannyboy11** [reply]
I agree with Ivan here - most plugins only require Player.canSee(targetPlayer), but I'm not sure whether other platforms besides Bukkit provide such an api, so this might still be useful.

**Reply by lokka30** [reply]
This API can help a lot with plugins that do vanishing with packets (not thru bukkit), and it also allows the requests to be done asynchronously. Doubt these queries even need to be done asynchronously but may as well make it possible.

**Reply by MrIvanPlays** [reply]
everything that has to do with the player character (potions, visibility, interactions) is not thread safe on any platform that is using the vanilla minecraft server as their base. vanishing a player async will result in an exception in bukkit.

**Reply by lokka30** [reply]
> everything that has to do with the player character (potions, visibility, interactions) is not thread safe on any platform that is using the vanilla minecraft server as their base. vanishing a player async will result in an exception in bukkit.

Nothing to do with potion effects etc - vanishing stops other players from receiving the *packets* that inform them that you exist on the server. Hides the player from in-game, tab list, player list count, player list in the motd hover message, can block vanished users from interacting with things like pressure plates, list goes on.

Treasury does not do anything regarding the interactions, visibility etc of vanished players. It only makes it very easy and central for a plugin to know if a player is vanished, and we can probably make it so plugins can make players vanished too. This is why I don't think we have any concerns over thread safety here.

All the vanish provider plugin needs to do is just cache the status. solved.

**Reply by MrIvanPlays** [reply]
we do have thread safety concerns. you can't vanish a player asynchronously. checking if he's vanished asynchronously may result in a race condition.

**Reply by lokka30** [reply]
> we do have thread safety concerns. you can't vanish a player asynchronously. checking if he's vanished asynchronously may result in a race condition.

Great point. Could the same be said with permissions too?
However, what if the vanish provider handled these requests on a single thread, still using the subscriber? Wouldn't that solve the race conditions issue?

**Reply by MrIvanPlays** [reply]
permissions can be thread safe if they are implemented properly.

**Comment by lokka30** [comment]
# Block Logging API

I am unsure what this API would do, but it involves interfacing with block logging plugins like CoreProtect and Prism.

## Likelihood:

Unsure.

## Features:

Unsure.

## Code ideas:

Unsure.

**Reply by MrIvanPlays** [reply]
I mean the only plugin that I am aware of that logs such data with interactions and stuff is CoreProtect. I don't think we need such API.

**Reply by Jikoo** [reply]
Prism and BigBrother also do block logging. There did used to be a unified API for them already, BlocksHub I believe, but it never really gained traction because it was limited by them, rather than the other way around.

CO's API is absolute crap for developers, would avoid it and leave people to their suffering tbh. They may have changed it to be less terrible, but last time I looked you basically had to assemble a query just like a user would in string form which CO would then re-parse. Not great options for restricting results either, pretty inflexible.

**Reply by lokka30** [reply]
I believe that this API is most likely unsuitable for Treasury. This is because I have no clue where it will be used. Don't block logging plugins just do everything internally? What plugins actually depend on the code of CoreProtect etc? @MrNemo64

**Reply by MrNemo64** [reply]
The idea for this came to my mind when thinking about the protections API. I couldn't think of a scenario where you would need it but thought that it was worth mentioning to see what you think. I wouldn't consider developing this further unless we see a necesity to do so

**Reply by Jikoo** [reply]
Yeah, I've had requests to support block loggers similarly to protection plugins for my use case, would see some niche use. Not sure if it's worth writing though given how different block logger APIs are, would need to research.

**Reply by MrIvanPlays** [reply]
I have no ideas how to write a block logger api too. We shall skip this until demand is my opinion.

**Reply by lokka30** [reply]
Should probably contact block logging plugin developers to see if they are interested in such an API.

**Reply by lokka30** [reply]
I don't believe we should develop such an API. What plugins need to use block logging plugin APIs? There are only a few block logging plugins that are popular at the moment, so I don't think a central API would help much.

**Comment by MrNemo64** [comment]
# 💬 Player Inputs API

## Likelihood:

Unsure.

## Features:

Allow a plugin to engage with a user in a conversation to gain inputs.

## Code ideas:
Provide an easy way to get (player) inputs in any way: (speaking from spigot) right now the ways of getting user input are:
  - Command arguments
  - Inventories (you have to make your system to handle the inventories)
  - Chat input, this may be with events (you have to make your system to handle the events) or the Conversation API.
 
A while ago this made me develop a [player chat inputs API](https://www.spigotmc.org/threads/player-chat-inputs.437565/)(which I've been trying to improve in the last week or so) and I've seen that it has also made @MrIvanPlays develop its own [Conversation API](https://www.spigotmc.org/threads/conversations-api.547930/). When I saw that at least 2 of us had this idea it made me think that having an API for this purpose should be considered.
The main problem and reason as to why we could need an API here is that if 2 plugins start an input process/conversation at the same time they would interfere. I also think that having an input API could bring some more attention to Treasury.
The API wouldn't be limited to chat inputs. My idea is to maybe separate the API into 3 parts:
1. The input receiver: this part of the API would be in charge of receiving inputs, via inventory, chat, commands, actions, clicks...
2. The input converter: this part of the API would be in charge of converting the inputs given by the receivers to the desired value (string(chat) to number, number(inventory slot) to boolean, etc)
3. The input users: this is the part of the API what would use the converted value and the one is mainly used by Treasury users

This way any input can be turned into any value. Also this is just an idea, not the way that it should be
This idea dosn't really fit into the provider/consumer idea but we already have the events and service APIs (I am aware that those exist to make Treasury more platform independent)
(It's late here so I won't answer messages until tomorrow)

**Reply by MrIvanPlays** [reply]
I didn't make it treasury specific api because its too cursed for a treasury api. And it would probably look like that here too.

**Reply by MrNemo64** [reply]
What do you mean with _cursed_? Hard to implement?

**Reply by MrIvanPlays** [reply]
Ever seen the base module?

**Reply by MrNemo64** [reply]
I have and I don't consider it that bad

**Reply by MrIvanPlays** [reply]
if you have anything better multi platform i'd like code. if you dont then this is not getting into treasury.

**Reply by MrNemo64** [reply]
I don't have time right now to write code but this may give an idea
```mermaid
classDiagram

    InputProcess --> InputSender
    InputProcess --> InputProcessor
    InputProcess --> InputListener
    InputProcess --> InputProcessState
    InputProcess --> InputProcessEndReason
    InputProcess --> InputProcessResponse
    InputProcessManager --> InputProcess

    class InputProcessManager {
        + startInputProcess(InputProcess~?, ?~ inputProcess)
        + Collection~InputProcess~ getQueuedInputProcessesFor(InputSender sender)
    }

    class InputProcess~InputType, ValueType~ {
        - InputProcessor~InputType, ValueType~ processor
        - InputListener~InputType~ listener
        - InputSender sender
        + boolean isRunning()
        + InputProcessState getInputProcessState()
        + void getInputProcessEndReason(InputProcessResponse~InputProcessEndReason~ subscriber)
        + void whenFinished(InputProcessResponse~ValueType~ subscriber)
    }

    class InputProcessEndReason {
        INPUT_ACCEPTED
        CANCELLED
        FORCE_CANCELLED
        COMPLETED_EXCEPTIONALLY
    }

    class InputProcessState {
        CREATED
        QUEUED
        RUNNING
        FINISHED
    }

    class InputSender {
        + UUID getUniqueId()
        + void sendMessage(String message)
    }

    class InputProcessor~InputType, ValueType~ {
        + boolean isValidInput(InputType input)
        + ValueType convertInput(InputType input)
    }

    class InputListener~InputType~ {
        + void receiveInput(InputType input)
    }
```

**Reply by MrIvanPlays** [reply]
thanks but my position here is no.

**Reply by lokka30** [reply]
The idea is great, but it is probably difficult or impossible to execute in a way that pertains with our high  standard for quality. I'm currently neutral for this API idea.

**Reply by lokka30** [reply]
I have thought about this further and I have come to the conclusion that although this API would be brilliant, unfortunately there are some key issues which prevent me from adding a conversations API.

1. Treasury is a library for service integrations. It adds specific APIs to make certain services such as economy plugin cores easier to integrate with. It however does not add utility APIs such as conversations, logging, database management, etc.
2. Low adoption. This does not sway me much but it is nevertheless something worth noting. Only a handful of plugins would benefit from such an API.

Thus, as we concluded with the Item Registry API, I believe this is worth being separate from Treasury, as much as it would be awesome to have both inside one central package.

**Comment by lokka30** [comment]
# ⚔️ Item Registry API

A registry for plugins like Oraxen and ItemsAdder, where they can register their custom items. Plugins can then, through only Treasury, interact with the items through our API.

## Alternatives:

I don't know of any existing item registry APIs, but I also haven't tried searching.

## Likelihood:

No idea.

## Features:
* Providers register items in Treasury's registry. Treasury stores the reference using a namespaced key
* Consumers can access a `RegisteredItem` by querying Treasury's registry using its namespaced key e.g. `oraxen:emerald_helmet`.
* the `RegisteredItem` objects make it easy to:
  * get the `ItemStack` of the registered item ***(platform dependence? discuss this)***
  * give a player the item
* Items can be registered and unregistered at any point
* It is expected that all items are unregistered by a given provider when their plugin disables. Treasury does not handle this
* ...more?

## Why?
There are several custom items plugins, and it is difficult for plugins to interact with them one by one especially considering some are premium. Usually we only need simple functions for the custom items anyways e.g. giving them to players and grabbing their item stack if we need to add it to mob drops on death or something.

## Needs discussion:
* How we are meant to implement this on other platforms
  * Proxies? Is it possible & feasible?

**Reply by lokka30** [reply]
Tagging those who may be interested: @MrIvanPlays @Jikoo @MrNemo64 

***

Credit to stumper66/PenalBuffalo for letting me know about this idea, and thanks to Machine-Maker for suggesting the use of NamespacedKeys for item registration.

**Reply by MrNemo64** [reply]
I see potential on this

**Reply by Jikoo** [reply]
Sounds neat, but as you brought up, platform independence is a big issue. I guess it wouldn't be too bad with per-platform registrations, but it's definitely harder to generify. In addition, it's for specific things that aren't present on certain implementations, i.e. Bungee does not have a concept of items. I wouldn't mind seeing this added, but I'm not sure how to do it right at a high level. I guess in those cases the registry would just always be empty?

<hr>

Re: premium plugins making things difficult:
While a Treasury API could certainly be helpful, the community refusing to bend over backwards and enable bad behavior from low-effort premium plugins is also important. Funnily enough, one of the projects I don't support I can't (without adding local files, at least) because it compiles against a binary of a closed-source premium custom item plugin.

I put this qualifier on my open offer for support of various plugins a while ago and my life has been much easier since:
> ### Adding Support
> [I rely] on plugins offering a public Maven-accessible API. If a plugin does not offer this, [I] cannot build against it. Contact the developer and ask them to either rectify this or implement their own hook.
> Anything requiring manual setup ruins the point of using a tool like Maven for dependency management. Maintaining support for plugins, [particularly plugins people paid someone else money for](https://www.appveyor.com/docs/how-to/secure-files/), requires a decent amount of additional manual work just to include (not counting updates) and reduces the ability of others to contribute.

**Reply by MrIvanPlays** [reply]
Very good idea, unfortunately unrealizable. If we're gonna add any api it should work on all platforms. This api will work only for bukkit currently ( and sponge in the future ). I'm not going to deal with the hassle of maintaining a complete wrapper api for items, especially considering the versions treasury aims to support ( < 1.13 is what im talking about ).

**Reply by lokka30** [reply]
Thank you for the feedback everyone! I agree with all of you. Perhaps I should create this as a separate plugin instead with only Bukkit and Sponge support?

**Reply by MrNemo64** [reply]
> Perhaps I should create this as a separate plugin instead with only Bukkit and Sponge support?

No one is going to stop you

**Reply by lokka30** [reply]
I would like to rekindle this conversation.

Treasury needs an item registry API - it will create a concrete API for this purpose, which seems yet to be fulfilled by other libraries. Therefore, this API will bring more plugins to support Treasury, i.e. the custom item adding plugins.

Previously, we concluded that the Item Registry API idea would not be feasible as proxies are unable to implement this. Though, I do not believe this is a valid reason to not create a useful API that can be used by all server platforms. I would also like to mention that the Regions API idea will also require a server-only implementation. It is well understood that proxies are limited in capability, so I don't see why anyone would be surprised that item management is not directly possible via proxy software.

To clarify, I am certainly against APIs that are exclusive against platforms in a particular group. For example, it would be a bad idea to create an API in Treasury that only supports Bukkit and BungeeCord, but not Sponge or Velocity.

I'd be grateful for the contributors and community to leave an opinion on this matter.

Tagging users who have already joined the conversation: @MrIvanPlays, @MrNemo64, @Jikoo

**Reply by lokka30** [reply]
I forgot to mention in my reply above that this API will only at most have a basic ItemStack implementation, if any. I was originally intending to not use ItemStack-like objects at all, though I see merit in adding such in.

**Reply by MrIvanPlays** [reply]
So you didn't get it at first? Okay, I'll tell you again and try to be more clear this time. 
1.13 bring a lot of changes to the items and materials. Due to the versions Treasury aims to support it is literally impossible to properly maintain such API, as it will be full of hacks. Sure, if such API is made the materials wouldn't be in an enum ( be in a registry ), but that is still impossible to maintain. And the mess will become even more incredible if we add any of the Via plugins in the mix, or ProtocolSupport. Sure, we mention that they are not supported, but do you think the average 1.8 server maintainer will care about your instructions/clear warnings not to use such plugins? 
Why can't proxies support items? Because it's not a proxy's job to do such stuff! A proxy only maintains the connection between a client and a server, in the meantime by adding a layer to make say global commands that work on every server the proxy is proxying. If a proxy adds anything else then the solid api of a proxy can break and then guess who's gonna be crying.

**Reply by lokka30** [reply]
> So you didn't get it at first? Okay, I'll tell you again and try to be more clear this time. 1.13 bring a lot of changes to the items and materials. Due to the versions Treasury aims to support it is literally impossible to properly maintain such API, as it will be full of hacks. Sure, if such API is made the materials wouldn't be in an enum ( be in a registry ), but that is still impossible to maintain. And the mess will become even more incredible if we add any of the Via plugins in the mix, or ProtocolSupport. Sure, we mention that they are not supported, but do you think the average 1.8 server maintainer will care about your instructions/clear warnings not to use such plugins? Why can't proxies support items? Because it's not a proxy's job to do such stuff! A proxy only maintains the connection between a client and a server, in the meantime by adding a layer to make say global commands that work on every server the proxy is proxying. If a proxy adds anything else then the solid api of a proxy can break and then guess who's gonna be crying.

Thank you for the reply -

Great point about legacy compatibility, I did not think of this.

I would like to add that I did not mention that proxies should support items, I reiterated that they don't and shouldn't. 

After further thought it seems that an item registry *must* use ItemStacks to be sufficiently useful for plugins. Thus it seems that a separate, 1.13+, Bukkit-only plugin needs to be developed to make an item registry possible. Back to the same conclusion as we had previously.

Again, thank you for the reply Ivan, much appreciated.

**Reply by MrNemo64** [reply]
I have no hard thoughts on this

**Comment by Jannyboy11** [comment]
> ## Alternatives:
> 
> I don't know of any existing item registry APIs, but I also haven't tried searching.

I know of one, but it doesn't seem widely used: https://polymart.org/resource/item-bridge.4

**Reply by lokka30** [reply]
> > ## Alternatives:
> > I don't know of any existing item registry APIs, but I also haven't tried searching.
> 
> I know of one, but it doesn't seem widely used: https://polymart.org/resource/item-bridge.4

I forgot to update this post, but I did start on an idea for [Registry](https://github.com/lokka30/Registry), I just haven't had the time to work on it yet. I've still yet to be able to pick up #180 and #167 😦

**Comment by thrailkillr** [comment]
# Punishments API
An API where advanced forms of punishment could possibly happen. ex: push a large cobblestone cube up a mountain to get your rights back

## Likelihood:
Unsure.
 
## Features:
PunishmentProvider: Like EconomyProvider but is used to deal punishments and their types.

## Code ideas:
PunishmentProvider, Punishment, PunishmentSubscriber (EconomySubscriber but called when the punishment is handed out/completed)

**Reply by lokka30** [reply]
Thanks for the suggestion @Hellscaped. More likely a punishments API would be very general and aim to cover mutes, warnings, bans, etc. I'm sure A248 would be very knowledgable about this if Ivan thinks it's a good idea to proceed with this idea.

**Reply by MrIvanPlays** [reply]
Noting that there are many ban plugins out there it'd be a good idea, however I can't see much use of it.

**Reply by lokka30** [reply]
( See OP's reply [here](https://github.com/ArcanePlugins/Treasury/discussions/128#discussioncomment-9536126) which was posted as a reply to the discussion rather than this existing thread ).

> > Noting that there are many ban plugins out there it'd be a good idea, however I can't see much use of it.
> 
> Most ban plugins have their punishment logic in them, this would separate the punishing from the punishment logic, making ban plugins able to have custom punishments (like the Sisyphus boulder idea) provided by additional plugins

So in other words, this would allow other plugins ("punishment consumers") to view, create, and edit punishments from ban plugins ("punishment providers")?

**Reply by thrailkillr** [reply]
> ( See OP's reply [here](https://github.com/ArcanePlugins/Treasury/discussions/128#discussioncomment-9536126) which was posted as a reply to the discussion rather than this existing thread ).
> 
> > > Noting that there are many ban plugins out there it'd be a good idea, however I can't see much use of it.
> > 
> > 
> > Most ban plugins have their punishment logic in them, this would separate the punishing from the punishment logic, making ban plugins able to have custom punishments (like the Sisyphus boulder idea) provided by additional plugins
> 
> So in other words, this would allow other plugins ("punishment consumers") to view, create, and edit punishments from ban plugins ("punishment providers")?

Oops, didn't realize I posted it as a reply to the discussion, sorry about that!

And yes, thats the idea in a nutshell.

**Comment by thrailkillr** [comment]
> Noting that there are many ban plugins out there it'd be a good idea, however I can't see much use of it.

Most ban plugins have their punishment logic in them, this would separate the punishing from the punishment logic, making ban plugins able to have custom punishments (like the Sisyphus boulder idea) provided by additional plugins
